hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/CustomFeeAssessmentStep.java

```diff
@@ -16,14 +16,261 @@
+import com.hedera.hapi.node.transaction.AssessedCustomFee;
+import com.hedera.node.app.service.token.ReadableTokenStore;
+import com.hedera.node.app.service.token.impl.handlers.transfer.customfees.AssessmentResult;
+import com.hedera.node.app.service.token.impl.handlers.transfer.customfees.CustomFeeAssessor;
+import com.hedera.node.app.service.token.impl.handlers.transfer.customfees.CustomFixedFeeAssessor;
+import com.hedera.node.app.service.token.impl.handlers.transfer.customfees.CustomFractionalFeeAssessor;
+import com.hedera.node.app.service.token.impl.handlers.transfer.customfees.CustomRoyaltyFeeAssessor;
+import com.hedera.node.app.service.token.impl.records.CryptoTransferRecordBuilder;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.config.data.LedgerConfig;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
 
 /**
- * Charges custom fees for the crypto transfer operation. This is yet to be implemented
+ * Charges custom fees for the crypto transfer operation.
+ * Custom fees can be a Fixed Fee (HBAR or HTS), Fractional Fee or Royalty Fee.
+ * Any fixed fees that are not self-denominated can trigger next level of custom fees assessment.
+ * Fractional fees and Royalty fees are not recursive.
+ * When assessing custom fees in this approach, we build list of transaction bodies that include assessed custom fees.
+ * We also build list of assessed custom fees to be added to the record.
+ * We do this in 2 steps:
+ * 1. Assess custom fees for the transaction body given as input (Level-0 Body)
+ * 2. If there are any fractional fees, adjust the assessed changes in the Level-0 Body
+ * 3. Any non-self denominated fixed (HBAR or HTS) fees, assess them and create Level-1 Body.
+ * But any self denominated fees will be adjusted in Level- 0 Body (since they can't trigger further custom fee charging.)
+ * 4.Any royalty fees which are not self denominated will be added to level-1 body.
  */
 public class CustomFeeAssessmentStep {
-    public CustomFeeAssessmentStep(final CryptoTransferTransactionBody op) {}
+    private final CryptoTransferTransactionBody op;
+    private final CustomFeeAssessor customFeeAssessor;
+    private int levelNum = 0;
+    private final HandleContext context;
+    private int totalBalanceChanges = 0;
+
+    public CustomFeeAssessmentStep(
+            @NonNull final CryptoTransferTransactionBody op, final TransferContextImpl transferContext) {
+        this.op = op;
+        this.context = transferContext.getHandleContext();
+        final var fixedFeeAssessor = new CustomFixedFeeAssessor();
+        final var fractionalFeeAssessor = new CustomFractionalFeeAssessor(fixedFeeAssessor);
+        final var royaltyFeeAssessor = new CustomRoyaltyFeeAssessor(fixedFeeAssessor);
+        customFeeAssessor = new CustomFeeAssessor(fixedFeeAssessor, fractionalFeeAssessor, royaltyFeeAssessor, op);
```

Comment:
Is there a reason this `CustomFeeAssessor` is constructed per-transaction, instead of being an "infrastructure" singleton like the `TransferContextImpl`?

Reply:
yeah that makes sense. Only reason I did this was to compute initialNftChanges.